test/jdk/java/foreign/TestStringEncoding.java

```diff
@@ -447,4 +457,61 @@ private static int strlen_byte(MemorySegment segment, long start) {
         throw new IllegalArgumentException("String too large");
     }
 
+    private static List<Arena> arenas() {
+        return Arrays.asList(
+                Arena.ofConfined(),          // Native memory
+                new HeapArena(byte.class),   // Heap memory backed by a byte array
+                new HeapArena(short.class),  // Heap memory backed by a short array
+                new HeapArena(int.class),    // Heap memory backed by an int array
+                new HeapArena(long.class));  // Heap memory backed by a long array
+    }
+
+    private static final class HeapArena implements Arena {
+
+        private final Class<?> type;
+        private final Arena delegate;
+
+        public HeapArena(Class<?> type) {
+            if (!type.isPrimitive()) {
+                throw new IllegalArgumentException(type.toString());
+            }
+            this.type = type;
+            this.delegate = Arena.ofConfined();
+        }
+
+        @Override
+        public MemorySegment allocate(long byteSize, long byteAlignment) {
+            var seg = switch (type) {
```

Comment:
Technically speaking, this is not an arena - as all returned segments have disjoint scopes. That said, you can try to create an heap-based arena using a slicing allocator? E.g. have HeapArena take a Java array - then wrap the array into a segment in the constructor, and wrap that segment into a SegmentAllocator. The Arena scope now is the scope of the heap segment. The close operation can be a no-op.